Reactions/rk64/reactor_rk64.cpp

```diff
@@ -0,0 +1,296 @@
+  amrex::Real* rY_src_in,
+  amrex::Real* rX_in,
+  amrex::Real* rX_src_in,
+  amrex::Real& dt_react,
+  amrex::Real& time,
+  int reactor_type,
+  int Ncells
+#ifdef AMREX_USE_GPU
+  ,
+  amrex::gpuStream_t stream
+#endif
+)
+{
+
+  amrex::Real time_init = time;
+  amrex::Real time_out = time + dt_react;
+  const amrex::Real tinyval = 1e-50;
+
+  // capture reactor type
+  int captured_reactor_type = reactor_type;
+  int captured_nsubsteps_guess = rk64_nsubsteps_guess;
+  int captured_nsubsteps_min = rk64_nsubsteps_min;
+  int captured_nsubsteps_max = rk64_nsubsteps_max;
+  amrex::Real captured_abstol = absTol;
+
+  int* nstepsvec;
+  nstepsvec = new int[Ncells]();
+
+  amrex::ParallelFor(Ncells, [=] AMREX_GPU_DEVICE(int icell) noexcept {
+    amrex::Real soln_reg[NUM_SPECIES + 1];
+    amrex::Real carryover_reg[NUM_SPECIES + 1];
+    amrex::Real error_reg[NUM_SPECIES + 1];
+    amrex::Real rhs[NUM_SPECIES + 1];
+    amrex::Real rYsrc[NUM_SPECIES];
+    amrex::Real dt_rk, dt_rk_min, dt_rk_max, change_factor;
+    amrex::Real current_time = time_init;
+    int neq = (NUM_SPECIES + 1);
+
+#include "rkparams.H"
+
+    for (int sp = 0; sp < neq; sp++) {
+      soln_reg[sp] = rY_in[icell * neq + sp];
+      carryover_reg[sp] = soln_reg[sp];
+      error_reg[sp] = 0.0;
+      rhs[sp] = 0.0;
+    }
+
+    dt_rk = dt_react / amrex::Real(captured_nsubsteps_guess);
+    dt_rk_min = dt_react / amrex::Real(captured_nsubsteps_max);
+    dt_rk_max = dt_react / amrex::Real(captured_nsubsteps_min);
+
+    current_time = time_init;
+    amrex::Real rhoe_init[] = {rX_in[icell]};
+    amrex::Real rhoesrc_ext[] = {rX_src_in[icell]};
+
+    for (int sp = 0; sp < NUM_SPECIES; sp++) {
+      rYsrc[sp] = rY_src_in[icell * NUM_SPECIES + sp];
+    }
+
+    int nsteps = 0;
+    while (current_time < time_out) {
+      for (int sp = 0; sp < neq; sp++) {
+        error_reg[sp] = 0.0;
+      }
+      for (int stage = 0; stage < nstages_rk64; stage++) {
+        fKernelSpec(
+          0, 1, current_time - time_init, captured_reactor_type, soln_reg, rhs,
+          rhoe_init, rhoesrc_ext, rYsrc);
+
+        for (int i = 0; i < neq; i++) {
+          error_reg[i] += err_rk64[stage] * dt_rk * rhs[i];
+          soln_reg[i] = carryover_reg[i] + alpha_rk64[stage] * dt_rk * rhs[i];
+          carryover_reg[i] = soln_reg[i] + beta_rk64[stage] * dt_rk * rhs[i];
+        }
+      }
+
+      current_time += dt_rk;
+      nsteps++;
+
+      amrex::Real max_err = tinyval;
+      for (int i = 0; i < neq; i++) {
+        if (fabs(error_reg[i]) > max_err) {
+          max_err = fabs(error_reg[i]);
+        }
+      }
+
+      if (max_err < captured_abstol) {
+        change_factor =
+          betaerr_rk64 * pow((captured_abstol / max_err), exp1_rk64);
+        dt_rk = amrex::min(dt_rk_max, dt_rk * change_factor);
+      } else {
+        change_factor =
+          betaerr_rk64 * pow((captured_abstol / max_err), exp2_rk64);
+        dt_rk = std::max(dt_rk_min, dt_rk * change_factor);
```

Comment:
These `std::max` need to be changed as well. We also need them to be `amrex::max<amrex::Real>` just in case we ever want to use single precision in the future.

Reply:
done